### PR TITLE
FORGOT PASSWORD

### PR DESCRIPTION
**Pull request by TarikVelic**
MAIN
 * Added forgot password function --------------------------------------------------------------------
 OTHER
 * <other msg> --------------------------------------------------------------------

### CODE REVIEW

File: controllers/forgotPassword.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const forgotPasswordService = require("../services/forgotPassword");
+const errors = require("../configuration/errors");
+
+exports.forgotPassword = async (req, res) => {
+  const { email } = req.body;
+
+  try {
+
+    const token = await forgotPasswordService.forgotPassword(email);
+    console.log(token);
```
Comment (becarselman) [inline controllers/forgotPassword.js]: remove console log

File: controllers/forgotPassword.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    res.status(200).json({ token });
+  } catch (error) {
+    console.error('Failed to send forgot password email:', error);
```
Comment (becarselman) [inline controllers/forgotPassword.js]: same

File: controllers/sendemail.js [hunk 3]
```diff
@@ -1,11 +1,12 @@
 const sendMailService = require('../services/sendemail')
 const errors = require("../configuration/errors")
+process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
```
Comment (becarselman) [inline controllers/sendemail.js]: what is this

File: services/forgotPassword.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    throw new Error("User not found");
+  }
+
+  const token = uuidv4();
+  
+  user.resetPasswordToken = token;
+  user.resetPasswordExpires = Date.now() + 3600000; // 1 hour
+
+  const templateName = "forgot-password";
+  const url = `${process.env.FRONTEND_URL}/reset-password/${token}`;
```
Comment (becarselman) [inline services/forgotPassword.js]: we have file for importing env vars

File: services/forgotPassword.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  const user = await User.findOne({ email: email });
+
+  if (!user) {
+    throw new Error("User not found");
+  }
+
+  const token = uuidv4();
+  
+  user.resetPasswordToken = token;
+  user.resetPasswordExpires = Date.now() + 3600000; // 1 hour
```
Comment (becarselman) [inline services/forgotPassword.js]: you can do this on token creation.

File: services/forgotPassword.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+const sendMailService = require("./sendemail");
+
+async function forgotPassword(email) {
+  const user = await User.findOne({ email: email });
+
+  if (!user) {
+    throw new Error("User not found");
+  }
+
+  const token = uuidv4();
```
Comment (becarselman) [inline services/forgotPassword.js]: token need to be jwt.

File: services/forgotPassword.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+const User = require("../db/models/User");
+const { v4: uuidv4 } = require('uuid');
+const sendMailService = require("./sendemail");
+
+async function forgotPassword(email) {
+  const user = await User.findOne({ email: email });
```
Comment (becarselman) [inline services/forgotPassword.js]: Move this to repository

File: services/forgotPassword.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  user.resetPasswordExpires = Date.now() + 3600000; // 1 hour
+
+  const templateName = "forgot-password";
+  const url = `${process.env.FRONTEND_URL}/reset-password/${token}`;
+  const context = {
+    resetLink: url
+  };
+
+  await sendMailService.sendEmail(email, templateName, context);
+  console.log('Forgot password email sent successfully.');
```
Comment (becarselman) [inline services/forgotPassword.js]: remove console.log

**Review by becarselman — CHANGES_REQUESTED** [review]

File: services/forgotPassword.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+const sendMailService = require("./sendemail");
+
+async function forgotPassword(email) {
+  const user = await User.findOne({ email: email });
+
+  if (!user) {
+    throw new Error("User not found");
+  }
+
+  const token = jwt.sign({ userId: user._id }, process.env.JWT_SECRET, { expiresIn: '1h' });
```
Comment (becarselman) [inline services/forgotPassword.js]: we have file for importing env vars.

File: package.json [hunk 10]
```diff
@@ -22,15 +22,17 @@
...
     "joi": "^17.9.2",
     "joi-password-complexity": "^5.1.0",
     "jsonwebtoken": "^9.0.0",
-    "mongodb": "^5.2.0",
+    "mongodb": "^5.4.0",
     "mongoose": "^7.0.3",
     "nodemailer": "^6.9.1",
     "nodemailer-express-handlebars": "^6.1.0",
     "nodeman": "^1.1.2",
+    "uuid": "^9.0.0",
```
Comment (becarselman) [inline package.json]: If we don't use uuid anymore we can remove it

File: package.json [hunk 11]
```diff
@@ -22,15 +22,17 @@
     "bcrypt": "^5.1.0",
     "cors": "^2.8.5",
     "dotenv": "^16.0.3",
+    "ejs": "^3.1.9",
     "express": "^4.18.2",
     "joi": "^17.9.2",
     "joi-password-complexity": "^5.1.0",
     "jsonwebtoken": "^9.0.0",
-    "mongodb": "^5.2.0",
+    "mongodb": "^5.4.0",
```
Comment (becarselman) [inline package.json]: why are you update version of this package

File: package.json [hunk 12]
```diff
@@ -22,15 +22,17 @@
     "bcrypt": "^5.1.0",
     "cors": "^2.8.5",
     "dotenv": "^16.0.3",
+    "ejs": "^3.1.9",
```
Comment (becarselman) [inline package.json]: If we dont use ejs remove it

**Review by becarselman — CHANGES_REQUESTED** [review]

File: repositories/user.js [hunk 13]
```diff
@@ -1,17 +1,25 @@
...
 
 async function createUser(userData) {
-    const { email, password, role } = { ...userData }
+  const { email, password, role } = { ...userData }
+
+  return await User.create({email: email, password: password, role: role})
+}
 
-    return await User.create({email: email, password: password, role: role})
+async function setResetPasswordToken(user) {
```
Comment (becarselman) [inline repositories/user.js]: Why you doing this in repository. Repository only need to communicate with db. Move it in some service file

File: repositories/user.js [hunk 14]
```diff
@@ -1,17 +1,25 @@
...
+  const token = jwt.sign({ userId: user._id }, process.env.JWT_SECRET, { expiresIn: '1h' });
+  user.resetPasswordToken = token;
+  user.resetPasswordExpires = Date.now() + 3600000; // 1 hour
```
Comment (becarselman) [inline repositories/user.js]: why you set expires date when we have already in jwt.

File: services/forgotPassword.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+  const user = await userRepository.getUserByEmail(email);
+
+  if (!user) {
+    throw new Error("User not found");
+  }
+
+  const token = await userRepository.setResetPasswordToken(user);
+
+  const templateName = "forgot-password";
+  const resetPasswordUrl = process.env.RESET_PASSWORD_URL;
```
Comment (becarselman) [inline services/forgotPassword.js]: move this process.env in configuration file

File: services/sendemail.js [hunk 16]
```diff
@@ -5,12 +5,15 @@ const fs = require("fs");
 
 require('dotenv').config()
```
Comment (becarselman) [inline services/sendemail.js]: I know this is not Your code but remove this line because we don't need

**Review by becarselman — CHANGES_REQUESTED** [review]

File: controllers/forgotPassword.js [hunk 17]
```diff
@@ -0,0 +1,15 @@
+const forgotPasswordService = require("../services/forgotPassword");
+const errors = require("../configuration/errors");
+
+exports.forgotPassword = async (req, res) => {
+  const { email } = req.body;
+
+  try {
+
```
Comment (becarselman) [inline controllers/forgotPassword.js:8]: Please pay attention to formatting in future

File: repositories/user.js [hunk 18]
```diff
@@ -1,17 +1,19 @@
-const User = require("../db/models/User")
-
+const User = require("../db/models/User");
+const jwt = require('jsonwebtoken');
```
Comment (becarselman) [inline repositories/user.js]: I think this is necessary import.

File: services/forgotPassword.js [hunk 19]
```diff
@@ -0,0 +1,35 @@
+const jwt = require("jsonwebtoken");
+const User = require("../db/models/User");
+const sendMailService = require("./sendemail");
+const userRepository = require("../repositories/user");
+const env = require("../configuration/env")
+
+async function setResetPasswordToken(user) {
```
Comment (becarselman) [inline services/forgotPassword.js]: same block of code we have in login service. Please create some helper function that will do this part(create token).

File: services/forgotPassword.js [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+  const token = jwt.sign({ userId: user._id }, process.env.JWT_SECRET, { expiresIn: '1h' });
+  return token;
+}
+
+async function forgotPassword(email) {
+  const user = await userRepository.getUserByEmail(email);
+
+  if (!user) {
+    throw new Error("User not found");
```
Comment (becarselman) [inline services/forgotPassword.js]: Why you are not reuse error message from errors file. I wrote this already.

File: configuration/env.js [hunk 21]
```diff
@@ -10,5 +10,6 @@ module.exports = {
     SMTP_PORT: process.env.SMTP_PORT,
     SMTP_SECURE: process.env.SMTP_SECURE === 'true',
     SMTP_USER: process.env.SMTP_USER,
-    SMTP_PASS: process.env.SMTP_PASS
+    SMTP_PASS: process.env.SMTP_PASS,
+    RESET_PASSWORD_URL: `${values.FRONTEND_URL}/reset-password/`
```
Comment (becarselman) [inline configuration/env.js]: You don't need to keep this here. Keep just frontend url. And i don't see here import of FRONTEND_URL. Did you test it. Is this work

**Review by becarselman — CHANGES_REQUESTED** [review]
I don't know how this working. You get email with resetPassword link. And I don't see where you create reset password function

File: controllers/resetPassword.js [hunk 22]
```diff
@@ -0,0 +1,15 @@
+const resetPasswordService = require('../services/resetPassword');
+const errors = require('../configuration/errors');
+
+exports.showResetPasswordForm = resetPasswordService.showResetPasswordForm;
+
+exports.resetPassword = async (req, res) => {
+  const { token } = req.params;
```
Comment (becarselman) [inline controllers/resetPassword.js]: we should get token from header authorization

File: email-templates/password-reset.success.hbs [hunk 23]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
```
Comment (becarselman) [inline email-templates/password-reset.success.hbs]: where we use this template

File: db/models/Token.js [hunk 24]
```diff
@@ -0,0 +1,23 @@
+const mongoose = require("mongoose");
+
+const tokenSchema = new mongoose.Schema({
+  _userId: {
+    type: mongoose.Schema.Types.ObjectId,
```
Comment (becarselman) [inline db/models/Token.js]: we don't need to store token in db

File: repositories/user.js [hunk 25]
```diff
@@ -1,17 +1,27 @@
...
```
Comment (becarselman) [inline repositories/user.js]: as far as I can see you resolve comment and do nothing 
image

File: services/resetPassword.js [hunk 26]
```diff
@@ -0,0 +1,42 @@
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+const userRepository = require('../repositories/user');
+const errors = require('../configuration/errors');
+
+async function showResetPasswordForm(req, res) {
```
Comment (becarselman) [inline services/resetPassword.js]: why we need this. we have frontend for displaying for. All you need is function that will take new password and token, verify user and update password. we don't need this on backend part.

File: services/resetPassword.js [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+  const token = req.params.token;
+  try {
+    const decodedToken = jwt.verify(token, process.env.JWT_SECRET);
+    const user = await userRepository.getUserById(decodedToken.userId);
+
+    if (!user) {
+      return res.status(404).send(errors.USER_NOT_FOUND);
```
Comment (becarselman) [inline services/resetPassword.js]: we use response only in controllers. and this is service

File: services/resetPassword.js [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+    }
+    res.render('resetPassword', { token });
```
Comment (becarselman) [inline services/resetPassword.js]: same

File: services/resetPassword.js [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+async function resetPassword(token, newPassword, confirmPassword) {
+  const decodedToken = jwt.verify(token, process.env.JWT_SECRET);
+  const user = await userRepository.getUserById(decodedToken.userId);
+
+  if (!user) {
+    throw new Error(errors.USER_NOT_FOUND);
+  }
+  if (decodedToken.exp < Date.now() / 1000) {
```
Comment (becarselman) [inline services/resetPassword.js]: I think verify will return false if token is expired. as well as if token is invalide for some other reason.

**Review by becarselman — CHANGES_REQUESTED** [review]

File: app.js [hunk 30]
```diff
@@ -5,21 +5,21 @@ const loginRoutes = require("./routes/login")
...
 
 const app = express()
 const port = env.PORT
 
 app.use(express.json())
-app.use(cors({
-  origin: FRONTEND_URL
-}))
-
+app.use(cors())
```
Comment (AntMarkoLovric) [inline app.js]: why did you remove frontend url?

File: configuration/errors.js [hunk 31]
```diff
@@ -3,10 +3,14 @@ module.exports = {
...
     EMAIL_TAKEN: "User with specified email already exists",
-    
+    PASSWORDS_DO_NOT_MATCH: "Passwords do not match",
     PASSWORD_NOT_PROVIDED: "Password not provided",
     INCORRECT_PASSWORD: "Incorrect password",
     FAILED_TO_SEND_EMAIL: "Failed to send email",
-    
-    USER_ROLE_NOT_FOUND: "User role not found"
+    USER_ROLE_NOT_FOUND: "User role not found",
+    USER_NOT_FOUND: "User not found",
```
Comment (AntMarkoLovric) [inline configuration/errors.js:11]: you added two of the same error messages, you could unite it into one and use it throughout the whole project

File: services/forgotPassword.js [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+  const user = await userRepository.getUserByEmail(email);
+
+  if (!user) {
+    throw new Error(errors.EMAIL_NOT_FOUND);
+  }
+
+  const token = await setResetPasswordToken(user);
+
+  const templateName = "forgot-password";
+  const resetPasswordUrl = `${process.env.FRONTEND_URL}/reset-password/`;
```
Comment (AntMarkoLovric) [inline services/forgotPassword.js]: extract "proces.env.FRONTEND_URL" into env.js, so you can reuse it, improves readability, and if you, at any point in development process need to change this, you will need to change in one place instead of changing in every place you use this variable

File: services/forgotPassword.js [hunk 33]
```diff
@@ -0,0 +1,35 @@
+const jwt = require("jsonwebtoken");
+const User = require("../db/models/User");
+const sendMailService = require("./sendemail");
+const userRepository = require("../repositories/user");
+const env = require("../configuration/env")
+
+async function setResetPasswordToken(user) {
+  const token = jwt.sign({ userId: user._id }, process.env.JWT_SECRET, { expiresIn: '1h' });
```
Comment (AntMarkoLovric) [inline services/forgotPassword.js]: move this line of code into a utils/helper.js function so you can use it wherever you need token creation

File: services/helper.js [hunk 34]
```diff
@@ -0,0 +1,10 @@
+const jwt = require("jsonwebtoken");
+const env = require("../configuration/env");
+
+function generateToken(payload, expiresIn) {
```
Comment (AntMarkoLovric) [inline services/helper.js]: you did this great, just move this into a utils folder

File: services/resetPassword.js [hunk 35]
```diff
@@ -0,0 +1,23 @@
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+const userRepository = require('../repositories/user');
+const errors = require('../configuration/errors');
+
+async function resetPassword(token, newPassword, confirmPassword) {
+  const decodedToken = jwt.verify(token, process.env.JWT_SECRET);
```
Comment (AntMarkoLovric) [inline services/resetPassword.js]: same as the comment above regarding moving this to env file

Comment (AntMarkoLovric) [reply services/resetPassword.js]: do we have some kind of validation if token is invalid?

**Review by AntMarkoLovric — CHANGES_REQUESTED** [review]

File: services/resetPassword.js (hunk 35 above)
Comment (TarikVelic) [reply services/resetPassword.js]: Screenshot_8

we get this kind of message

File: services/forgotPassword.js [hunk 36]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+async function forgotPassword(email) {
+  const user = await userRepository.getUserByEmail(email);
+
+  if (!user) {
+    throw new Error(errors.EMAIL_NOT_FOUND);
+  }
+
+  const token = await setResetPasswordToken(user);
```
Comment (becarselman) [inline services/forgotPassword.js]: we don't need extra function for this. We already have generateToken function use it here

File: services/resetPassword.js [hunk 37]
```diff
@@ -0,0 +1,24 @@
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+const userRepository = require('../repositories/user');
+const errors = require('../configuration/errors');
+const env = require("../configuration/env")
+
+async function resetPassword(token, newPassword, confirmPassword) {
```
Comment (becarselman) [inline services/resetPassword.js]: we don't need to take confirm password because we will validate that on frontend.

File: services/resetPassword.js [hunk 38]
```diff
@@ -0,0 +1,24 @@
...
+  const decodedToken = jwt.verify(token, env.JWT_SECRET);
```
Comment (becarselman) [inline services/resetPassword.js]: create helper function for this

**Review by becarselman — CHANGES_REQUESTED** [review]

File: utils/helper.js [hunk 39]
```diff
@@ -0,0 +1,18 @@
+const jwt = require("jsonwebtoken");
+const env = require("../configuration/env");
+const errors = require('../configuration/errors');
+
+function generateToken(payload, expiresIn) {
```
Comment (becarselman) [inline utils/helper.js]: please use arrow functions

File: services/resetPassword.js [hunk 40]
```diff
@@ -0,0 +1,23 @@
...
+const decodedToken = verifyJwtToken(token, env.JWT_SECRET);
+const user = await userRepository.getUserById(decodedToken.userId);
+
+if (!user) {
+throw new Error(errors.USER_NOT_FOUND);
+}
+
+const salt = await bcrypt.genSalt(10);
+const hashedPassword = await bcrypt.hash(newPassword, salt);
+await userRepository.updateUserPassword(user._id, hashedPassword);
```
Comment (becarselman) [inline services/resetPassword.js]: return something

**Review by becarselman — APPROVED** [review]

**Review by ElvaRivle — APPROVED** [review]
